Pattern location group update

br648 opened the pull request:
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [x] The description lists all applicable issues this PR seeks to resolve
- [x] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR includes updates to handle pattern location groups. This closely follows pattern locations. It has only been tested as far as unit tests and not from the UI. E2E testing will be carried out once the approriate UI updates are available.

Commit:
refactor(Pattern location group addition): First pass at including pattern location groups

binh-dam-ibigroup commented on src/main/java/com/conveyal/gtfs/loader/JdbcGtfsSnapshotter.java:
```diff
@@ -103,10 +103,11 @@ public SnapshotResult copyTables() {
             result.locationShapes = copy(Table.LOCATION_SHAPES, true);
             result.routes = copy(Table.ROUTES, true);
-            // FIXME: Find some place to store errors encountered on copy for patterns and pattern stops.
+            // FIXME: Find some place to store errors encountered on copy various patterns.
```
"on copy" => "after copying"

binh-dam-ibigroup commented on src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java:
```diff
@@ -173,10 +174,9 @@ public String update(Integer id, String json, boolean autoCommit) throws SQLExce
-                            !childEntities.isArray())
+                    if ((referencingTable.name.equals(Table.PATTERN_LOCATION.name) ||
+                        referencingTable.name.equals(Table.PATTERN_LOCATION_GROUP.name)) &&
+                        (childEntities == null || childEntities.isNull() || !childEntities.isArray())
```
Can `ArrayUtils.isEmpty()` replace this line?

br648 replied:
It doesn't look like it. The checks have to be made within the context of a `JsonNode`. This line is repeated eight lines below so have extracted into a separate method.

binh-dam-ibigroup replied:
Ok, thanks for checking.

binh-dam-ibigroup commented on src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java:
```diff
@@ -312,11 +312,18 @@ public int normalizeStopTimesForPattern(int id, int beginWithSequence) throws SQ
+            JDBCTableReader<PatternLocationGroup> patternLocationGroups = new JDBCTableReader(
+                Table.PATTERN_LOCATION_GROUP,
+                dataSource,
+                tablePrefix + ".",
```
I'm not the biggest fan of appending the dot here. I'm gonna note an issue for refactoring that later.

br648 replied:
... so no action for me here then?

binh-dam-ibigroup replied:
No action. That change is not relevant to this PR and could be breaking too.

binh-dam-ibigroup commented on src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java:
```diff
@@ -905,19 +949,22 @@ private int updateStopTimesForPatternStop(PatternStop patternStop, int previousT
+     * @return the travel and dwell time added by this pattern.
      * @throws SQLException
      */
-    private int updateStopTimesForPatternLocation(PatternLocation patternLocation, int previousTravelTime, String tripId)
```
I'd say keep this shorthand method, name it `updateStopTimes`, and make it call `updateStopTimesForPatternLocationOrPatternLocationGroup` with the correct params, and add a similar method that accepts a `PatternLocationGroup` to simplify the corresponding method calls.

binh-dam-ibigroup commented on src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java:
```diff
@@ -741,9 +767,12 @@ private String updateChildTable(
+                // Patterns must only increase, not increment.
                 boolean valuesAreAscending =
-                    (!Table.PATTERN_LOCATION.name.equals(subTable.name) && !Table.PATTERN_STOP.name.equals(subTable.name))
+                    (!Table.PATTERN_LOCATION.name.equals(subTable.name) &&
+                        !Table.PATTERN_STOP.name.equals(subTable.name) &&
+                        !Table.PATTERN_LOCATION_GROUP.name.equals(subTable.name)
+                    )
```
Replace with `!isPatternTable`.

br648 replied:
Good spot, thanks.

binh-dam-ibigroup commented on src/main/java/com/conveyal/gtfs/validator/PatternFinderValidator.java:
```diff
@@ -109,20 +115,27 @@ public void complete(ValidationResult validationResult) {
+            PreparedStatement insertPatternStatement = connection.prepareStatement(patternsTable.generateInsertSql(true));
+            PreparedStatement insertPatternStopStatement = connection.prepareStatement(patternStopsTable.generateInsertSql(true));
+            PreparedStatement insertPatternLocationStatement = connection.prepareStatement(patternLocationsTable.generateInsertSql(true));
+            PreparedStatement insertPatternLocationGroupStatement = connection.prepareStatement(patternLocationGroupsTable.generateInsertSql(true));
```
These statements should be closed.

binh-dam-ibigroup commented on src/main/java/com/conveyal/gtfs/validator/PatternFinderValidator.java:
```diff
@@ -172,8 +185,7 @@ public void complete(ValidationResult validationResult) {
                     }
-
-                    if (stopById.containsKey(stopOrLocationId)) {
+                    if (stopById.containsKey(stopOrLocationIdOrLocationGroupId)) {
```
Is the below the implementation for `setStatementParameters` for `PatternStop`? Same question for the other blocks that call `addLocationOrLocationGroup`.

br648 replied:
The first block covers pattern stops. The other blocks cover locations and location groups. Because locations and location groups are very similar, they are covered in the same method (addLocationOrLocationGroup). What is your concern with this section of code? I'm not sure I follow.

binh-dam-ibigroup replied:
Is `setStatementParameters` intended to contain that code for the respective pattern classes?

br648 replied:
Yes, although a couple of params were not in the correct order. Thanks for the heads up, sort of :-)

binh-dam-ibigroup replied:
My question was, could the `setIntParameters` and others be moved to method `setStatementParameters` in classes `PatternStop`, `PatternLocation` and `PatternLocationGroup`? That method is presently empty in all these classes.

br648 replied:
ah right sorry! Yep, refactored.

binh-dam-ibigroup commented on src/test/java/com/conveyal/gtfs/loader/JDBCTableWriterTest.java:
```diff
@@ -1306,11 +1403,11 @@ public void canReconcilePatterns(PatternArguments patternArguments)
                 entry.getKey(),
                 stopSequence,
-                (cumulativeTravelTime += entry.getValue() == PatternReconciliation.PatternType.LOCATION
+                (cumulativeTravelTime += (entry.getValue() == PatternReconciliation.PatternType.LOCATION || entry.getValue() == PatternReconciliation.PatternType.LOCATION_GROUP)
```
Maybe extract the condition into a var.

binh-dam-ibigroup reviewed (changes requested):
LGTM, mostly refactoring remarks/questions.

Commit:
refactor(Addressed PR feedback): Addressed PR feedback

binh-dam-ibigroup commented on src/main/java/com/conveyal/gtfs/loader/JdbcTableWriter.java:
```diff
@@ -871,25 +863,17 @@ private void updatePatternFrequencies(PatternReconciliation reconciliation) thro
-                        patternLocation.pattern_id,
-                        patternLocation.stop_sequence,
+                    updateStopTimes(
+                        reconciliation.getPatternLocation(genericStop.referenceId),
```
Yep, this is much better.

binh-dam-ibigroup reviewed (approved):
Looks good so far. I am still wondering about moving the `setIntParameter` and others to the applicable pattern classes, but I'm not holding that against this PR.

Commit:
refactor(Addressed PR feedback): Refactored pattern finder validator and moved insert statements int

Commit:
refactor(Multi location group update): Update to handle creation of multi location groups in one sub

Commit:
refactor(Rollback to the previous commit): Multi location group update is no longer required

This commit will match the commit hash of 6c471f99b61bb066b2b6ac3939fb6ba5799d85ae

Commit:
refactor(Unpacking and packing of multi location groups): Bespoke handling of location groups with m

Commit:
refactor(Table.java): Using string literal for location group file name

The location group table reference hasn't be created at this point which was causing all sorts of
fun.

miles-grant-ibigroup committed:
Merge pull request #355 from conveyal/multi-loc-grp

Multiple Location Groups

miles-grant-ibigroup committed:
refactor: add location groups to graphql

miles-grant-ibigroup committed:
refactor(GraphQLGtfsSchema): attempt to repair pattern location groups

Commit:
refactor(Unit test updates and fixed bug): Updated graphql unit tests to address graphql enhancement

Commit:
refactor(Updated pattern cascade delete): cascade delete now includes pattern locations and location

miles-grant-ibigroup commented:
Having finished the front-end companion PR for this, I would say this is now ready for merge/second review

Commit:
refactor(Code formatting): Aligned method parameters

binh-dam-ibigroup commented on src/main/java/com/conveyal/gtfs/loader/JdbcGtfsExporter.java:
```diff
@@ -471,4 +498,20 @@ public static int writeLocationsToFile(
+        zipOutputStream.putNextEntry(new ZipEntry(LOCATION_GROUPS_FILE_NAME));
+        // Create and use PrintWriter, but don't close. This is done when the zip entry is closed.
+        PrintWriter p = new PrintWriter(zipOutputStream);
+        p.println(LocationGroup.packLocationGroups(locationGroups));
```
The combination of `printLn` and the addition of a line return in `LocationGroup.getCsvRow` (via `packLocationGroup`) is causing a double blank line at the end of the exported `location_groups.txt`, which, when re-imported, causes a "wrong number of columns" error on the first blank line (Similar to the issue handled by https://github.com/ibi-group/datatools-server/pull/473).

br648 replied:
Thanks for this. Have updated and tested output.

binh-dam-ibigroup reviewed (changes requested):
One comment regarding the double blank line.

Commit:
refactor(JdbcGtfsExporter.java): Removed redundant new line

binh-dam-ibigroup reviewed (approved):
Approving for now, thanks for the changes. The unknown to check when UI code is fixed is regarding saving location group timetables.
